packages/core/src/browser/preferences/preference-validation-service.ts

```diff
@@ -0,0 +1,278 @@
+                case 'number':
+                    return this.validateNumber(key, value, schema);
+                case 'object':
+                    return this.validateObject(key, value, schema);
+                case 'string':
+                    return this.validateString(key, value, schema);
+                default:
+                    this.validateNever(schema.type, key);
+                    return value;
+            }
+        } catch (e) {
+            console.error('Encountered an error while validating', key, 'with value', value, 'against schema', schema, e);
+            return value;
+        }
+    }
+
+    protected getSchema(name: string, overrideIdentifier?: string): ValidatablePreference | undefined {
+        const combinedSchema = this.schemaProvider.getCombinedSchema().properties;
+        if (combinedSchema[name]) {
+            return combinedSchema[name];
+        }
+        const baseName = this.languageOverrideService.overriddenPreferenceName(name)?.preferenceName;
+        return baseName !== undefined ? combinedSchema[baseName] : undefined;
+    }
+
+    protected validateMultiple(key: string, value: JSONValue, schema: ValidatablePreference & { type: JsonType[] }): JSONValue {
+        const validation: ValidatablePreference = deepClone(schema);
+        const candidate = this.mapValidators(key, value, (function* (this: PreferenceValidationService): Iterable<ValueValidator> {
+            for (const type of schema.type) {
+                validation.type = type as JsonType;
+                yield toValidate => this.validateBySchema(key, toValidate, validation);
+            }
+        }).bind(this)());
+        if (candidate !== value && (schema.default !== undefined || schema.defaultValue !== undefined)) {
+            const configuredDefault = this.getDefaultFromSchema(schema);
+            return this.validateMultiple(key, configuredDefault, { ...schema, default: undefined, defaultValue: undefined });
+        }
+        return candidate;
+    }
+
+    protected validateAnyOf(key: string, value: JSONValue, schema: ValidatablePreference & { anyOf: ValidatablePreference[] }): JSONValue {
+        const candidate = this.mapValidators(key, value, (function* (this: PreferenceValidationService): Iterable<ValueValidator> {
+            for (const option of schema.anyOf) {
+                yield toValidate => this.validateBySchema(key, toValidate, option);
+            }
+        }).bind(this)());
+        if (candidate !== value && (schema.default !== undefined || schema.defaultValue !== undefined)) {
+            const configuredDefault = this.getDefaultFromSchema(schema);
+            return this.validateAnyOf(key, configuredDefault, { ...schema, default: undefined, defaultValue: undefined });
+        }
+        return candidate;
+    }
+
+    protected mapValidators(key: string, value: JSONValue, validators: Iterable<(value: JSONValue) => JSONValue>): JSONValue {
```

Comment:
Some of the validators require `key` because they make recursive calls back to the entry point or log failures, so I decided to add `key` to all of them so that later modifications that do either of those things aren't breaking, but if that seems excessively cautious, it can be removed where it isn't used.